common/lib/client/rest.ts

```diff
@@ -0,0 +1,224 @@
+		}
+		const optionsObj = Defaults.objectifyOptions(options);
+
+		if(optionsObj.log) {
+			Logger.setLog(optionsObj.log.level, optionsObj.log.handler);
+		}
+		Logger.logAction(Logger.LOG_MICRO, 'Rest()', 'initialized with clientOptions ' + Utils.inspect(options));
+
+		const normalOptions = this.options = Defaults.normaliseOptions(optionsObj);
+
+		/* process options */
+		if(normalOptions.key) {
+			const keyMatch = normalOptions.key.match(/^([^:\s]+):([^:.\s]+)$/);
+			if(!keyMatch) {
+				const msg = 'invalid key parameter';
+				Logger.logAction(Logger.LOG_ERROR, 'Rest()', msg);
+				throw new Error(msg);
+			}
+			normalOptions.keyName = keyMatch[1];
+			normalOptions.keySecret = keyMatch[2];
+		}
+
+		if('clientId' in normalOptions) {
+			if(!(typeof(normalOptions.clientId) === 'string' || normalOptions.clientId === null))
+				throw new ErrorInfo('clientId must be either a string or null', 40012, 400);
+			else if(normalOptions.clientId === '*')
+				throw new ErrorInfo('Can’t use "*" as a clientId as that string is reserved. (To change the default token request behaviour to use a wildcard clientId, use {defaultTokenParams: {clientId: "*"}})', 40012, 400);
+		}
+
+		Logger.logAction(Logger.LOG_MINOR, 'Rest()', 'started; version = ' + Defaults.version);
+
+		this.baseUri = this.authority = function(host) { return Defaults.getHttpScheme(normalOptions) + host + ':' + Defaults.getPort(normalOptions, false); };
+		this._currentFallback = null;
+
+		this.serverTimeOffset = null;
+		this.auth = new Auth(this, normalOptions);
+		this.channels = new Channels(this);
+		this.push = new Push(this);
+	}
+
+	stats(params: RequestParams, callback: StandardCallback<PaginatedResult<Stats>>): Promise<PaginatedResult<Stats>> | void {
+		/* params and callback are optional; see if params contains the callback */
+		if(callback === undefined) {
+			if(typeof(params) == 'function') {
+				callback = params;
+				params = null;
+			} else {
+				if(this.options.promises) {
+					return Utils.promisify(this, 'stats', arguments) as Promise<PaginatedResult<Stats>>;
+				}
+				callback = noop;
+			}
+		}
+		const headers = Utils.defaultGetHeaders(),
+			format = this.options.useBinaryProtocol ? Utils.Format.msgpack : Utils.Format.json,
+			envelope = Http.supportsLinkHeaders ? undefined : format;
+
+		if(this.options.headers)
+			Utils.mixin(headers, this.options.headers);
+
+		(new PaginatedResource(this, '/stats', headers, envelope, function(body: unknown, headers: Record<string, string>, unpacked?: boolean) {
+			const statsValues = (unpacked ? body : JSON.parse(body as string));
+			for(let i = 0; i < statsValues.length; i++) statsValues[i] = Stats.fromValues(statsValues[i]);
+			return statsValues;
+		})).get(params as Record<string, string>, callback);
+	}
+
+	time(params: RequestParams | StandardCallback<number>, callback?: StandardCallback<number>): Promise<number> | void {
+		/* params and callback are optional; see if params contains the callback */
```

Comment:
Intriguing. Have we got an issue in the backlog to improve this API? Well, assuming you think it _does_ need improving, however seeing the preamble required in this method's implementation I am thinking that if an interface like this can't be cleanly expressed in TypeScript then it needs a refactor. 🤔

Reply:
So this is really a problem caused by us using the old NodeJS callback-style async API. The fact that the callback is always the last argument (rather than in a fixed position) means that most of the public API methods in the library have to do the same argument normalisation dance. The only way we can really solve this is through a huge breaking API change so it's not really an option in v1.x versions of the library. Once we can release v2 then as part of that I would like for us to move to a modern ES6 promise-based async API which would make all of this arg normalisation stuff redundant and allow us to remove a lot of lib code.

Reply:
Thanks. Great explanation. Appreciated. 😁